common/to.etc.alg/src/main/java/to/etc/util/DiagnosticUtil.java

```diff
@@ -0,0 +1,144 @@
+package to.etc.util;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Handy util to do profiling on duration spent in certain code blocks.
+ * Call logInit to initialize in memory logging that is stored in thread locale.
+ * Call logStart with given string key-name on start of code block to measure. Call logStop with given same key-name at end of code block, to stop measuring.
+ * It would lof start time plus duration kept in precise nano seconds. It also allows for easy measuring in nested code paths as loops, it automatically
+ * keeps records on each separate run and also produces an overall record for given key-name.
+ *
+ * Example (with given 'test overall duration' key around loop, and 'test iteration duration' key inside loop:
+ * test overall duration started at 13:29:26.743=157
+ * test iteration duration (total count 3) started at 13:29:26.759=41
+ * test iteration duration_#1 started at 13:29:26.759=11
+ * test iteration duration_#2 started at 13:29:26.774=15
+ * test iteration duration_#3 started at 13:29:26.790=15
+ *
+ * To get the Map with durations report, call report method at the end, with specified time unit for reporting. That also unregister logging from thread locale
+ * automatically.
+ *
+ * If initLog is not called, called log methods are ignored, and report returns null result.
+ */
+public class DiagnosticUtil {
+
+	private static ThreadLocal<Map<String, Object>> diagnostic = new ThreadLocal<>();
+	private static ThreadLocal<Boolean> diagnosticEnabled = new ThreadLocal<>();
+
+	private static final String SUB_MAP = "_subMap";
+
+	private static final SimpleDateFormat DF = new SimpleDateFormat("HH:mm:ss.SSS");
```

Comment:
These are not reentrant, so you cannot use the instance from multiple threads. And this code WILL be called from multiple threads... You have no other choice than to just allocate one when needed.